src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/4.diff.sql

```diff
@@ -447,6 +477,132 @@ AS
+--
+-- RETURN VALUE
+--     The tag error fields and the corresponding instance UIDs.
+/***************************************************************************************/
+CREATE OR ALTER PROCEDURE dbo.GetExtendedQueryTagErrors (@tagPath VARCHAR(64))
+AS
+    SET NOCOUNT     ON
+    SET XACT_ABORT  ON
+    BEGIN TRANSACTION
+
+        DECLARE @tagKey     INT
+        SELECT @tagKey = TagKey
+        FROM dbo.ExtendedQueryTag WITH(HOLDLOCK)
+        WHERE dbo.ExtendedQueryTag.TagPath = @tagPath
+
+        -- Check existence
+        IF (@@ROWCOUNT = 0)
+            THROW 50404, 'extended query tag not found', 1 
+
+        SELECT
+            TagKey,
+            ErrorCode,
+            CreatedTime,
+            StudyInstanceUid,
+            SeriesInstanceUid,
+            SopInstanceUid
+        FROM dbo.ExtendedQueryTagError AS XQTE
+        INNER JOIN dbo.Instance AS instance ON XQTE.Watermark = instance.Watermark
+        WHERE TagKey = @tagKey
+
+    COMMIT TRANSACTION
+GO
+
+/***************************************************************************************/
+-- STORED PROCEDURE
+--    DeleteExtendedQueryTagErrors
+--
+-- DESCRIPTION
+--    Delete errors for specified extended query tag
+--
+-- PARAMETERS
+--     @tagPath
+--         * The extended query tag path
+/***************************************************************************************/
+CREATE OR ALTER PROCEDURE dbo.DeleteExtendedQueryTagErrors (@tagPath VARCHAR(64))
+AS
+    SET NOCOUNT     ON
+    SET XACT_ABORT  ON
+    BEGIN TRANSACTION
+
+        DECLARE @tagKey INT
+
+        SELECT @tagKey = TagKey
+        FROM dbo.ExtendedQueryTag WITH(XLOCK)
+        WHERE dbo.ExtendedQueryTag.TagPath = @tagPath
+
+        -- Check existence
+        IF (@@ROWCOUNT = 0)
+            THROW 50404, 'extended query tag not found', 1 
+
+        DELETE FROM dbo.ExtendedQueryTagError
+        WHERE TagKey = @tagKey
+
+    COMMIT TRANSACTION
+GO
+
+/***************************************************************************************/
+-- STORED PROCEDURE
+--     AddExtendedQueryTagError
+--
+-- DESCRIPTION
+--    Adds an Extended Query Tag Error or Updates it if exists.
+--
+-- PARAMETERS
+--     @tagKey
+--         * The related extended query tag's key
+--     @errorCode
+--         * The error code
+--     @watermark
+--         * The watermark
+--     @createdTime
+--         * The time the error was created
+--
+-- RETURN VALUE
+--     The tag key of the error added.
+/***************************************************************************************/
+CREATE OR ALTER PROCEDURE dbo.AddExtendedQueryTagError (
+    @tagKey INT,
+    @errorCode INT,
+    @watermark BIGINT,
+    @createdTime DATETIME2(7)
+)
+AS
+    SET NOCOUNT     ON
+    SET XACT_ABORT  ON
+    BEGIN TRANSACTION
+
+        --Check if tag exists
+        IF NOT EXISTS (SELECT * FROM dbo.ExtendedQueryTag WITH (HOLDLOCK) WHERE TagKey = @tagKey)
```

Comment:
do we want to check if the DICOM instance exists as well?

Reply:
I check tag because it is the key for my relationship between XQT table and XQTerror table.
I think checking instance exist, adding another hold lock is unnecessary.